### PR TITLE
Sign in

### PR DESCRIPTION
**Pull request by KaotharReda**
When user sign up request: name, email, userType (should be either BROKER, BUYER, ADMIN or RENTER, you could use a radiobutton for that input) and password. 
The system check the name if it already exist registration fail
When the user sign in request only name and password

### CODE REVIEW

File: server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java
Comment (huy2272) [inline server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java:1]: Do we still need `Buyer` class when we already have `User`?

File: server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java [hunk 1]
```diff
@@ -4,30 +4,45 @@
...
     }
 
-    public User saveUser(User user) {
+    public Optional<Broker> getBroker(String id){
+        return brokerRepository.findById(id);
+    }
+
+    public User signUp(User user) {
+        // Check if user already exists
+        if (userRepository.findByName(user.getName()) != null){
```
Comment (huy2272) [inline server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java:31]: Nice! Love that you are thinking about this edge case

File: server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java [hunk 2]
```diff
@@ -4,30 +4,45 @@
...
 
-    public Optional<Broker> getBroker(String id){
-        return brokerRepository.findById(id);
+    public User signIn(String name, String password) {
+        User user = userRepository.findByName(name);
+
+        if (user != null && user.getPassword().equals(password)) {
+            return user;
+        } else {
+            return null; // Authentication failed
```
Comment (huy2272) [inline server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java:43]: We should return `403 Forbidden Error` here.

**Review by huy2272 — CHANGES_REQUESTED** [review]

File: server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java
Comment (KaotharReda) [reply server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java:1]: I could delete still it got no use

File: server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java (hunk 2 above)
Comment (KaotharReda) [reply server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java:43]: yeah this give 500, how can we make it return 403

Comment (KaotharReda) [reply server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java:43]: ohh not even it 200 but return no body

File: server/real_estate/src/main/java/com/realestate/real_estate/controllers/UsersController.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    private UserService userService;
+
+    @PostMapping("/sign-up")
+    public User signUp(@RequestBody User user) {
+        return userService.signUp(user);
+    }
+
+    @PostMapping("/sign-in")
+    public User signIn(@RequestParam String name, @RequestParam String password) {
+        return userService.signIn(name, password);
```
Comment (huy2272) [inline server/real_estate/src/main/java/com/realestate/real_estate/controllers/UsersController.java:23]: @KaotharReda you can check if the method returns null or not here.
If null we return http.Forbidden.
Something like this

```
if (userService.signIn(name, password)) {
            return ResponseEntity.ok().build();
        }
        return new ResponseEntity<>(null, HttpStatus.FORBIDDEN);
```

File: server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java
Comment (huy2272) [reply server/real_estate/src/main/java/com/realestate/real_estate/repos/Buyer.java:1]: if it's useless, it's better to delete

File: server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java (hunk 2 above)
Comment (huy2272) [reply server/real_estate/src/main/java/com/realestate/real_estate/services/UserService.java:43]: check my comment below

**Review by huy2272 — APPROVED** [review]

**Pull request merged**